Camera Mode Button

tomaszrybakiewicz opened the pull request:
Closes [#1434](https://github.com/mapbox/navigation-sdks/issues/1434)

### Description

_libnavui-maps_
- implemented `MapboxCameraModeButton` 

_libnavui-util_
- updated `ExtendableButtonHelper` to allow measurement functions injection

_qa-test-app_
- added `DropInButtonsActivity` for `MapboxCameraModeButton` testing

### Screenshots or Gifs

Screen recording from `qa-test-app` testing MapboxCameraModeButton with default and custom style.

https://user-images.githubusercontent.com/2678039/153982716-9b75fbf7-d52f-428c-a230-dd1e0964db06.mp4

abhishek1508 commented on libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/camera/view/MapboxCameraModeButton.kt:
```diff
@@ -0,0 +1,185 @@
+    /**
+     * Camera state that this button represents.
+     */
+    var cameraState: NavigationCameraState = NavigationCameraState.FOLLOWING
```
I don't think we should maintain any state in this `View`. Looking at other UI components, we have veered away from storing states as much as possible for 2 reasons:
- None of our views are capable of retaining states upon config change.
- This adds more logic to the view thereby preventing us from making the view as dumb as possible.

However, there are exceptions where more complicated views do maintain a state in one way or another. I agree that it's tempting to do this to provide ease of integration to the developers, but now that we have DropInUI and it exactly provides a mechanism to do this state management, we might want to keep it there. Past MVP the idea is to move individual `ViewBinders` and `UIComponents` to their respective module and that's where we can do this.

tomaszrybakiewicz replied:
This is entirely up to you. This View has been modeled to act more like a [CheckBox](https://developer.android.com/reference/android/widget/CheckBox) then a simple Button.

I will remove `cameraState` and update both `setState` methods to be no longer idempotent.

abhishek1508 commented on libnavui-maps/src/main/res/drawable/mapbox_bg_button.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android">
+    <solid android:color="#ffffff" />
```
Instead of using hardcoded colors, can we use them from `colors.xml`? Same applies to all the other colors in this entire PR

abhishek1508 commented on libnavui-maps/src/main/res/drawable/mapbox_bg_button.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android">
+    <solid android:color="#ffffff" />
+    <stroke android:color="#CDD0D0" android:width="2dp" />
```
Border color and width should also be a customizable attribute.

tomaszrybakiewicz replied:
Developers can customize any background attributes by providing their own background drawable.

abhishek1508 commented on libnavui-maps/src/main/res/values/attrs.xml:
```diff
@@ -17,4 +17,19 @@
+        <!-- Follow mode text -->
+        <attr name="cameraModeButtonOverviewText" format="string" localization="suggested" />
+        <!-- Overview mode text -->
+        <attr name="cameraModeButtonFollowText" format="string" localization="suggested" />
```
Let's add `borderColor` and `borderWidth`

abhishek1508 commented on qa-test-app/src/main/res/drawable-v24/bg_custom_button.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Can we move this file to `drawable` rather than in `drawable-v24`?

tomaszrybakiewicz commented on libnavui-maps/src/main/res/drawable/mapbox_bg_button.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
```
@abhishek1508 I am extremely confused here. The new buttons are rectangular with 10px border radius. https://www.figma.com/file/6bXLQGCUaYb7DdJsNL3e0Z/DS-Navigation?node-id=3334%3A39612
Am I working with wrong designs?

abhishek1508 replied:
No you are correct that they are rectangular by design. All I am saying is that we should make them that way in the DropInUI by providing a different background. Since this PR is adding a standalone UI component, to me it makes sense to have them look the same as other floating buttons; in this case circular. Later, in DropInUI module we can use this rectangular background and apply it to all the buttons

tomaszrybakiewicz replied:
Can you send me a link to Figma designs I should use?

tomaszrybakiewicz commented on libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/camera/view/MapboxCameraModeButton.kt:
```diff
@@ -0,0 +1,185 @@
+
+    private fun updateIconDrawable() {
+        if (isInFollowMode(cameraState)) {
+            iconImage.setImageResource(followingIconResId)
```
What does this button represent?
1. Current NavigationCameraState (current implementation)
2. Next NavigationCameraState

With the latter option, what text should appear when we expand the view?

abhishek1508 replied:
This button represents next `NavigationCameraState`. If the camera is in following and the button shows overview icon, when pressed the camera should transition to overview and button icon should change to following with the text as "Following"

tomaszrybakiewicz commented:
@abhishek1508 Can you take another look? I've updated Button implementation as you suggested. 

I didn't change button background resource yet. I need Figma designs for this.
 
Here's a screen recording of updated button.

https://user-images.githubusercontent.com/2678039/154145595-55c9c840-05d0-4468-a9f4-bbc29b34aa13.mp4

abhishek1508 commented:
> I didn't change button background resource yet. I need Figma designs for this.

You can use the same resource as we are using in `MapboxSoundButton` or `MapboxRecenterButton` @tomaszrybakiewicz

tomaszrybakiewicz commented:
@abhishek1508 Updated. Please review.

tomaszrybakiewicz commented on libnavui-maps/src/main/res/values/strings.xml:
```diff
@@ -2,4 +2,5 @@
 <resources xmlns:tools="http://schemas.android.com/tools">
     <string name="mapbox_recenter" tools:ignore="MissingTranslation">Recenter</string>
     <string name="mapbox_route_overview" tools:ignore="MissingTranslation">Overview</string>
+    <string name="mapbox_route_follow" tools:ignore="MissingTranslation">Follow</string>
```
@abhishek1508 Not sure if the "Follow" is the right text we should use here. 
Also we need to add all translations for it. Any idea how to do it?

abhishek1508 replied:
@tomaszrybakiewicz Let's use `Following`. Also, I wouldn't worry about translations for now, because we generally get out translations from contributors. I and Minh are working separately to get more contributors to contribute to translations on Transifex.

abhishek1508 commented on libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/camera/view/MapboxCameraModeButton.kt:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    private fun getText(state: NavigationCameraState): String? =
+        if (isInFollowMode(state)) followingText else overviewText
```
```suggestion
        if (isInFollowMode(state)) overviewText else followingText
```

abhishek1508 commented on libnavui-maps/src/main/res/layout/mapbox_camera_mode_button_layout.xml:
```diff
@@ -0,0 +1,41 @@
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="wrap_content"
+    android:layout_height="wrap_content"
+    tools:background="@drawable/mapbox_bg_button"
```
I don't see `mapbox_bg_button` anywhere. Can we replace it with the correct one?

tomaszrybakiewicz replied:
Good catch. Forgot to update this one.

abhishek1508 commented on libnavui-maps/src/main/res/drawable/mapbox_ic_camera_follow.xml:
```diff
@@ -0,0 +1,14 @@
+    android:viewportHeight="32">
+  <path
+      android:pathData="M15.3334,8.0001C14.9652,8.0001 14.6667,7.6684 14.6667,7.2593V1.4075C14.6667,0.9984 14.9652,0.6667 15.3334,0.6667H16.6667C17.0349,0.6667 17.3334,0.9984 17.3334,1.4075V7.2593C17.3334,7.6684 17.0349,8.0001 16.6667,8.0001H15.3334Z"
+      android:fillColor="#3C3E3E"
```
Can we use the colors from colors.xml? Let's not worry about the colors in figma design because they are gonna come through NSS. Once available, we will change the colors for all views at once.

I would suggest to add a new property to `libnavui-maps/res/values/colors.xml`

```
<resources>
    <color name="mapbox_camera_mode_drawable_color">@color/colorOnSurface</color>
</resources>
```

tomaszrybakiewicz replied:
Icon colors are applied via `cameraModeButtonIconTint` attribute. I've updated `MapboxStyleCameraModeButton` style to apply `@color/colorOnSurface` tint by default.

tomaszrybakiewicz commented:
@abhishek1508 Applied your suggestions. Can you give it another go?

tomaszrybakiewicz commented:
@abhishek1508 I've rebased on top of latest `main`, moved `MapboxCameraModeButton` to _libnavui-dropin_ module and updated its background to match figma designs as we discussed.
I kept `mapbox_route_follow` string in _libnavui-maps_ module to match `mapbox_route_overview`.
Please let me know if there is anything else you would like to change.

abhishek1508 reviewed (approved):
There are certain UI components in Figma designs which don't exist as standalone components today. These include:
- Toggle button to toggle between camera following and overview
- Start and end navigation button
- Recenter and sound button with a different design

Since, we don't have enough time to build these as standalone components first and then resue them in DropInUI, we are adding them directly in DropInUI. There are advantages to this approach:

- Adding these buttons as standalone widgets doesn't make sense at this moment as we are pretty sure that they might break in terms of styling and theming with the rollout of NSS. So technically we would be exposing them as standalone widgets just to deprecate them sooner.
- It enables us to add these buttons simply not worrying too much about NSS and theming system. Since DropInUI is still experimental, we would be easily able to change these buttons and finally extract them out as a standalone widget.

Approving this PR so we can move forward with DropInUI and figure out the best approach to make them standalone.
